core/src/stored/backends/crud_storage.cc

```diff
@@ -0,0 +1,362 @@
+
+   This program is Free Software; you can redistribute it and/or
+   modify it under the terms of version three of the GNU Affero General Public
+   License as published by the Free Software Foundation and included
+   in the file LICENSE.
+
+   This program is distributed in the hope that it will be useful, but
+   WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+   Affero General Public License for more details.
+
+   You should have received a copy of the GNU Affero General Public License
+   along with this program; if not, write to the Free Software
+   Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+   02110-1301, USA.
+*/
+#include "include/bareos.h"
+#include <fmt/format.h>
+#include "crud_storage.h"
+#include "lib/berrno.h"
+#include "lib/bpipe.h"
+#include "lib/bstringlist.h"
+#include "stored/stored_conf.h"
+#include "stored/stored_globals.h"
+#include <sys/stat.h>
+#include <cctype>
+#if defined(HAVE_WIN32)
+#  include <shlwapi.h>  // for PathIsRelative()
+#endif
+
+namespace {
+constexpr int debug_info = 110;
+constexpr int debug_trace = 130;
+
+bool path_is_relative(const std::string& path)
+{
+#if defined(HAVE_WIN32)
+  return PathIsRelative(path.c_str());
+#else
+  return path[0] != '/';
+#endif
+}
+
+class BPipeHandle {
+  Bpipe* bpipe{nullptr};
+
+ public:
+  BPipeHandle(const char* prog,
+              int wait,
+              const char* mode,
+              const std::unordered_map<std::string, std::string>& env_vars = {})
+      : bpipe(OpenBpipe(prog, wait, mode, true, env_vars))
+  {
+    if (!bpipe) { throw std::system_error(ENOENT, std::generic_category()); }
+  }
+  BPipeHandle(const BPipeHandle&) = delete;
+  BPipeHandle& operator=(const BPipeHandle&) = delete;
+  BPipeHandle(BPipeHandle&&) = delete;
+  BPipeHandle& operator=(BPipeHandle&&) = delete;
+
+  ~BPipeHandle()
+  {
+    if (bpipe) { CloseBpipe(bpipe); }
+  }
+  FILE* getReadFd() { return bpipe->rfd; }
+  FILE* getWriteFd() { return bpipe->wfd; }
+  std::string getOutput()
+  {
+    close_write();
+    std::string output;
+    char iobuf[1024];
+    while (!feof(bpipe->rfd)) {
+      size_t rsize = fread(iobuf, 1, 1024, bpipe->rfd);
+      if (rsize > 0 && !ferror(bpipe->rfd)) { output.append(iobuf, rsize); }
+    }
+    return output;
+  }
+  void reset_timeout() { TimerKeepalive(*bpipe->timer_id); }
+  bool timed_out() { return bpipe->timer_id && bpipe->timer_id->killed; }
+  void close_write()
+  {
+    ASSERT(bpipe);
+    CloseWpipe(bpipe);
+  }
+  int close()
+  {
+    ASSERT(bpipe);
+    int ret = CloseBpipe(bpipe) & ~b_errno_exit;
+
+    if (ret & b_errno_signal) {
+      ret &= ~b_errno_signal;
+      ret *= -1;
+    }
+
+    bpipe = nullptr;
+    return ret;
+  }
+};
+
+bool is_valid_env_name(const std::string& name)
+{
+  // according to IEEE Std 1003.1-2017 names should be
+  // non-empty, alphanumeric and not starting with a digit.
```

Comment:
I dont have access to this, but if it actually says this, then we should not allow `_`.

Reply:
> Environment variable names used by the utilities in the Shell and Utilities volume of IEEE Std 1003.1-2001 consist solely of uppercase letters, digits, and the '_' (underscore) from the characters defined in Portable Character Set and do not begin with a digit

I did find this however.  Did they remove the uppercase requirement in 2017 ?  We should still update the comment!

Reply:
`std::isalnum` depends on locale.  Maybe we should just check for `A-Z0-9_` ourselves.

Reply:
So we have a new standard from this year: https://pubs.opengroup.org/onlinepubs/9799919799/basedefs/V1_chap08.html
The requirement is 
> names shall not contain any bytes that have the encoded value of the character '='.

However, there is also the following guarantee for variables names used:
> Environment variable names used by the utilities in the Shell and Utilities volume of POSIX.1-2024 consist solely of uppercase letters, digits, and the <underscore> ('_') from the characters defined in [Portable Character Set](https://pubs.opengroup.org/onlinepubs/9799919799/basedefs/V1_chap06.html#tagtcjh_3) and do not begin with a digit.

Later it says:
> The name space of environment variable names containing lowercase letters is reserved for applications. Applications can define any environment variables with names from this name space without modifying the behavior of the standard utilities. 

So my conclusion is: `[A-Z_][A-Z0-9_]+` is to configure behaviour of system utilities, `[A-Za-z_][A-Za-z0-9_]+` containing at least one lower-case letter is for configuring application behaviour.

I'll update the comment to point to the latest standard and use an ASCII-version of `isalnum()`.